doc/sphinx/language/gallina-specification-language.rst

```diff
@@ -392,8 +465,61 @@ stands for :n:`let @ident := fun {+ @binder} => @term in @term’`.
 
 .. index:: match ... with ...
 
-Definition by case analysis
----------------------------
+Definition by cases: match
+--------------------------
+
+.. insertgram term_match record_pattern
+
+.. productionlist:: coq
+   term_match : match `case_item_list_comma` `return_type_opt` with `or_opt` `eqn_list_or_opt` end
+   case_item_list_comma : `case_item_list_comma` , `case_item`
+                        : `case_item`
+   return_type_opt : `return_type`
+                   : `empty`
+   as_return_type : `as_name_opt` `return_type`
+                  : `empty`
+   return_type : return `term100`
+   case_item : `term100` `as_name_opt` `in_opt`
+   as_name_opt : as `name`
+               : `empty`
+   in_opt : in `pattern`
+          : `empty`
+   or_opt : |
+          : `empty`
+   eqn_list_or_opt : `eqn_list_or`
+                   : `empty`
+   eqn_list_or : `eqn_list_or` | `eqn`
+               : `eqn`
+   eqn : `pattern_list_comma_list_or` => `term`
+   pattern_list_comma_list_or : `pattern_list_comma_list_or` | `pattern_list_comma`
+                              : `pattern_list_comma`
+   pattern_list_comma : `pattern_list_comma` , `pattern`
+                      : `pattern`
+   pattern : `pattern10` : `term`
+           : `pattern10`
+   pattern10 : `pattern1` as `name`
+             : `pattern1_list`
+             : @ `qualid` `pattern1_list_opt`
+             : `pattern1`
+   pattern1_list : `pattern1_list` `pattern1`
+                 : `pattern1`
+   pattern1_list_opt : `pattern1_list`
+                     : `empty`
+   pattern1 : `pattern0` % `ident`
+            : `pattern0`
+   pattern0 : `qualid`
+            : {| `record_patterns` |}
+            : _
+            : ( `pattern` )
+            : ( `pattern` | `pattern_list_or` )
+            : `numeral`
+            : `string`
+   pattern_list_or : `pattern_list_or` | `pattern`
+                   : `pattern`
+   record_patterns : `record_pattern` ; `record_patterns`
+                   : `record_pattern`
+                   : `empty`
```

Comment:
Yes. Right recursion in the grammar is a bit of an anti-pattern.  Below are some comments on this that I wrote some time ago.  Should we change this in the mlgs (and remove the superfluous production) too?

---------
I noticed that coqpp defines lists of tokens using right-recursion, such as "symbol SEMICOLON symbols" instead of the left-recursive "symbols SEMICOLON symbol".

```
symbols:
| symbol { [$1] }
| symbol SEMICOLON symbols { $1 :: $3 }
;
```

A right-recursive form may get a shift-reduce conflict that doesn't occur with the left-recursive form.  For example, the last line here would get a conflict when it sees a SEMICOLON if it were written right-recursively:

```
doc_gram_symbols_opt:
| { [] }
| doc_gram_symbols { $1 }
| doc_gram_symbols SEMICOLON { $1 }
;

doc_gram_symbols:
| doc_gram_symbol { [$1] }
| doc_gram_symbols SEMICOLON doc_gram_symbol { $1 @ [$3] }
;
```

At least on the C and Java versions of yacc, left recursion is recommended because it uses less space on the parser stack.  Right recursion accumulates the entire list on the parser stack which could overflow for a very large list when left recursion doesn't have that problem.

Of course, left-recursion needs an action such as "{ $1 @ [$3] }", which is N^2 over the entire list, though in practice probably unimportant.

Reply:
I don't know yacc and LALR well. In camlp5, there are built-in `LIST0`, `LIST1`, `LIST0SEP` and `LIST1SEP`. They are implemented tail-recursively, so that they don't push on the stack.

In your example, the whole of `doc_gram_symbols_opt` could be rewritten in LL style so as to avoid the conflict, the same way as e.g. `record_fields` is defined in `g_constr.mlg`.

So, I'm not a specialist, but the left- vs right-recursion issue seems more related to the LR vs LL parsing question (see e.g. this [comment](https://en.wikipedia.org/wiki/LL_parser#Special_Case:_Left_Recursion) about left recursion in LL parsers). The `LIST` operators of camlp5 are internally implemented using right recursion, and camlp5 itself is actually a LL-parser.

Reply:
I think LALR vs LL doesn't matter.  If your grammar is right-recursive, you get a parse tree like `(rps (rp ";" (rp ";" (rp ";" empty))))`.  Since actions are applied bottom-up, the parser has to save state for all the unapplied actions until it gets to `(rp ";" empty)`

With left-recursion, you get a parse tree like `(rps (((empty";" rp) ";" rp) ";" rp))` and the parser only needs to save state for one unapplied action.

I wonder why the grammar doesn't use `LIST0` here.

My main point is that writing something with right-recursion may prevent the parser from parsing some inputs that left-recursion would permit.  Whether the stack the could overflow depends on how our parser is written.  With yacc, the parser stack is a fixed-length array, so it can overflow for long inputs with right-recursion.  In contrast, the actions build up a list, which is only limited by the memory available to the process.

> (see e.g. this comment about left recursion in LL parsers)

I think the comment means that it is *possible* to get such conflicts for certain grammars, not that left recursion should not be used.

But a bit of a tangent.

Reply:
> I wonder why the grammar doesn't use `LIST0` here.

Because the syntax for record accepts an optional `;` at the end. As far as I can judge, this cannot be simulated with a `LIST0` nor a `LIST0 ... SEP ...`. One would need an extra variant of `LIST0`.

Anyway, is the long-term goal to use `prodn`? If yes, the question about left or right recursion is going to be moot, or do I miss something?